tests/unittests/Foundation/NSDecimalNumberTests.mm

```diff
@@ -0,0 +1,269 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <TestFramework.h>
+#import <Foundation/Foundation.h>
+
+static void assertArrayEqual(unsigned short* a, unsigned short* b, int len) {
+    int length = len - 1;
+    while (length >= 0) {
+        ASSERT_EQ(a[length], b[length]);
+        length--;
+    }
+}
+
+OSX_DISABLED_TEST(NSDecimalNumber, decimalNumberWithMantissa) {
+    unsigned short mantissa[] = { 1, 5, 9, 9 };
+    NSDecimalNumber* price = [NSDecimalNumber decimalNumberWithMantissa:1599 exponent:-2 isNegative:NO];
+
+    NSDecimal decimal = [price decimalValue];
+    ASSERT_EQ(4, decimal._length);
+    ASSERT_EQ(-2, decimal._exponent);
+    ASSERT_EQ(0, decimal._isNegative);
+    assertArrayEqual(mantissa, decimal._mantissa, decimal._length);
+    ASSERT_EQ(15.99, [price doubleValue]);
+
+    price = [NSDecimalNumber decimalNumberWithMantissa:159900 exponent:-2 isNegative:NO];
+    decimal = [price decimalValue];
+    ASSERT_EQ(4, decimal._length);
+    ASSERT_EQ(0, decimal._exponent);
+    ASSERT_EQ(0, decimal._isNegative);
+    assertArrayEqual(mantissa, decimal._mantissa, decimal._length);
+    ASSERT_EQ(1599.0, [price doubleValue]);
+
+    price = [NSDecimalNumber decimalNumberWithMantissa:159900 exponent:-2 isNegative:YES];
+    decimal = [price decimalValue];
+    ASSERT_EQ(4, decimal._length);
+    ASSERT_EQ(0, decimal._exponent);
+    ASSERT_EQ(1, decimal._isNegative);
+    assertArrayEqual(mantissa, decimal._mantissa, decimal._length);
+    ASSERT_EQ(-1599.0, [price doubleValue]);
+}
+
+TEST(NSDecimalNumber, decimalNumberAddition) {
+    NSDecimalNumber* d1 = [NSDecimalNumber decimalNumberWithMantissa:1599 exponent:-2 isNegative:NO];
```

Comment:
You also need:

- negative positive mixed test
- test values with much larger mantissas (greater than 2^32, so it takes up more than 2 shorts' worth)